WGCNA Part 3: DE and heatmaps (#363)
AlexsLemonade/refinebio-examples: 04-advanced-topics/network-analysis_rnaseq_01_wgcna.Rmd

  dplyr::filter(rowSums(.) >= 50)
```

Another thing we need to do is make sure our main experimental group label is set up. 
In this case `refinebio_treatment` has two groups: `pre-adt` and `post-adt`. 

> We need this to be a factor, but we also need `pre-adt` as the first level. 

Reviewer: Clarifying the conversion to a factor (we don't _need_ it, but it is nice to do)
```suggestion
To keep these two treatments in logical (rather than alphabetical) order, we will convert this to a factor with `pre-adt` as the first level. 
```

```{r}
metadata <- metadata %>%

>   dplyr::mutate(refinebio_treatment = factor(refinebio_treatment,

Reviewer: This is a place where I might make a new variable name for simplicity, but I don't feel strongly about it.

Author: If you didn't feel strongly, I didn't do it.

fit <- limma::eBayes(fit)
```

Apply multiple testing correction and obtain stats in a data frame. 

```{r}
# Apply multiple testing correction and obtain stats
stats_df <- limma::topTable(fit, number = nrow(module_eigengenes_t)) %>%
  tibble::rownames_to_column("module")
```

Let's arrange the results by the smallest adjusted p values. 

```{r}
stats_df %>%
  dplyr::arrange(adj.P.Val)
```

Module 52 seems to be the most differentially expressed across `refinebio_treatment` groups. 
Now we can do some investigation into this module. 

## Let's make plot of module 52

As a sanity check, let's use `ggplot` to see what module 52's eigengene looks like between treatment groups. 

```{r}
module_eigengenes %>%
  tibble::rownames_to_column("accession_code") %>%
  dplyr::inner_join(metadata %>%
    dplyr::select(refinebio_accession_code, refinebio_treatment),
  by = c("accession_code" = "refinebio_accession_code")
  ) %>%
  ggplot2::ggplot(ggplot2::aes(
    x = refinebio_treatment,
    y = ME52,
    color = refinebio_treatment
  )) +
  ggforce::geom_sina() +
  ggplot2::theme_classic()
```

This makes sense!

## What genes are a part of module 52?

> If you want to know which of your genes make up a modules, you can look at the `$colors` slot. 

Author: Agreed. I hate it.

This is a named list which associates the genes with the module they are a part of. 
We can turn this into a data frame for handy use. 

```{r}
gene_module_key <- tibble::enframe(bwnet$colors, name = "gene", value = "module") %>%
  # Let's add the `ME` part so its more clear what these numbers are and it matches elsewhere
  dplyr::mutate(module = paste0("ME", module))
```

Now we can find what genes are a part of module 52. 

```{r}
gene_module_key %>%
  dplyr::filter(module == "ME52")
```

Let's save this gene to module key to a TSV file for future use. 

```{r}
readr::write_tsv(gene_module_key,
  file = file.path("results", "SRP133573_wgcna_gene_to_module.tsv")
)
```

## Make a custom heatmap function

We will make a heatmap that summarizes our differentially expressed module.
Because we will make a couple of these, it makes sense to make a custom function for making this heatmap. 

```{r}

> make_module_heatmap <- function(module_name = "ME52",

Author: Whoops this was leftover from me testing this.

        dplyr::pull(module_name)
    ) %>%
    # Store sample
    tibble::column_to_rownames("refinebio_accession_code") %>%
    # Arrange by treatment
    dplyr::arrange(refinebio_treatment)

  # Create the ComplexHeatmap column annotation object
  col_annot <- ComplexHeatmap::HeatmapAnnotation(
    # Supply treatment labels
    refinebio_treatment = col_annot_df$refinebio_treatment,
    # Add annotation barplot
    module_eigengene = ComplexHeatmap::anno_barplot(col_annot_df$module_eigengene),
    # Pick colors for each experimental group in refinebio_treatment
    col = list(refinebio_treatment = c("post-adt" = "#f1a340", "pre-adt" = "#998ec3"))
  )

  # Get a vector of the Ensembl gene ides that correspond to this module
  module_genes <- gene_module_key_df %>%
    dplyr::filter(module == module_name) %>%
    dplyr::pull(gene)

  # Set up the gene expression data frame
  mod_mat <- expression_mat %>%
    t() %>%
    as.data.frame() %>%
    # Only keep genes from this module
    dplyr::filter(rownames(.) %in% module_genes) %>%
    # Order the samples to match col_annot_df
    dplyr::select(rownames(col_annot_df)) %>%
    # Data needs to be a matrix
    as.matrix()

>   if (standardize_genes) {

Author: It's advanced topics; maybe it's fine??

    # Genes are on very different scales; we will standardize them

>     mod_mat <- t(apply(mod_mat, 1, scale))

Author: Okey dokey.

    # Create a color function based on standardized scale
    color_func <- circlize::colorRamp2(
      c(-1, 0, 1),
      c("#67a9cf", "#f7f7f7", "#ef8a62")
    )
  } else {
    # Create a color function based on standardized scale
    color_func <- circlize::colorRamp2(
      c(
        quantile(mod_mat, 0.05),
        median(mod_mat),

>         quantile(mod_mat, 0.95)

Author: I think having the option to easily shut it off is nice.

      ),
      c("#67a9cf", "#f7f7f7", "#ef8a62")
    )
  }

  # Plot on a heatmap
  heatmap <- ComplexHeatmap::Heatmap(mod_mat,
    name = module_name,
    # Supply color function
    col = color_func,
    # Supply column annotation
    bottom_annotation = col_annot,
    # We don't want to cluster samples
    cluster_columns = FALSE,
    # We don't need to show sample or gene labels
    show_row_names = FALSE,
    show_column_names = FALSE
  )

  # Return heatmap
  return(heatmap)
}
```

## Make module heatmaps

Let's try out the custom heatmap function with module 52 (our most differentially expressed module).

```{r}
mod_52_heatmap <- make_module_heatmap(module_name = "ME52")

# Print out the plot

> mod_52_heatmap

Reviewer: I don't love that the legends are side by side instead of one above the other, but the fix seems like it adds way too much complexity, so I would rather live with this.

```

> We can save this plot to PDF.

Reviewer: Similarly to the first plot, I would add some text here to aid in interpretation. What should a reader be getting out of this plot?

  # metadata_df: a data frame with refinebio_accession_code and refinebio_treatment
  #              as columns. Default is `metadata`.
  # gene_module_key: a data.frame indicating what genes are a part of what modules. Default is `gene_module_key`.
  # module_eigengenes: a sample x eigengene data.frame with samples as rownames. Default is `module_eigengenes`.
  #
  # Returns:
  # A heatmap of expression matrix for a module's genes, with a barplot of the
  # eigengene expression for that module.

  # Set up the module eigengene with its refinebio_accession_code
  module_eigengene <- module_eigengenes_df %>%
    dplyr::select(module_name) %>%
    tibble::rownames_to_column("refinebio_accession_code")

  # Set up column annotation from metadata
  col_annot_df <- metadata_df %>%
    # Only select the treatment and sample ID columns
    dplyr::select(refinebio_accession_code, refinebio_treatment) %>%
    # Add on the eigengene expression by joining with sample IDs
    dplyr::inner_join(module_eigengene, by = "refinebio_accession_code") %>%
    # Store sample
    tibble::column_to_rownames("refinebio_accession_code") %>%
    # Arrange by treatment

>     dplyr::arrange(refinebio_treatment)

Reviewer: Can you test something here? What if you include `refinebio_title` arrange by this as well I think this should put replicates together? Or see if just arranging  by accession number helps?
i.e. `dplyr::arrange(refinebio_treatment, refinebio_accession)`, or `dplyr::arrange(refinebio_treatment, refinebio_title)`

Author: Turns out replicates are in the order of accession IDs so I've used that for ordering.

  # Create the ComplexHeatmap column annotation object
  col_annot <- ComplexHeatmap::HeatmapAnnotation(
    # Supply treatment labels
    refinebio_treatment = col_annot_df$refinebio_treatment,
    # Add annotation barplot
    module_eigengene = ComplexHeatmap::anno_barplot(dplyr::select(col_annot_df, module_name)),
    # Pick colors for each experimental group in refinebio_treatment
    col = list(refinebio_treatment = c("post-adt" = "#f1a340", "pre-adt" = "#998ec3"))
  )

>   # Get a vector of the Ensembl gene ides that correspond to this module

Reviewer: ```suggestion
  # Get a vector of the Ensembl gene ids that correspond to this module
```

    # Order the samples to match col_annot_df
    dplyr::select(rownames(col_annot_df)) %>%
    # Data needs to be a matrix
    as.matrix()

  mod_mat <- mod_mat %>%
    # Scale can work on matrices, but it does it by column so we will need to
    # transpose first
    t() %>%
    scale() %>%
    # And now we need to transpose back
    t()

  # Create a color function based on standardized scale
  color_func <- circlize::colorRamp2(
    c(-2, 0, 2),
    c("#67a9cf", "#f7f7f7", "#ef8a62")
  )

  # Plot on a heatmap
  heatmap <- ComplexHeatmap::Heatmap(mod_mat,
    name = module_name,
    # Supply color function
    col = color_func,
    # Supply column annotation
    bottom_annotation = col_annot,
    # We don't want to cluster samples
    cluster_columns = FALSE,
    # We don't need to show sample or gene labels
    show_row_names = FALSE,
    show_column_names = FALSE
  )

  # Return heatmap
  return(heatmap)
}
```

## Make module heatmaps

Let's try out the custom heatmap function with module 52 (our most differentially expressed module).

```{r}
mod_52_heatmap <- make_module_heatmap(module_name = "ME52")

# Print out the plot
mod_52_heatmap
```

> From the barplot portion of our plot, we can see `post-adt` samples have higher values for this eigengene for module 52.

Reviewer: Spacing for better rendering. Comment looks good!
```suggestion

From the barplot portion of our plot, we can see `post-adt` samples have higher values for this eigengene for module 52.
```

In the heatmap portion, we can see how the individual genes that make up module 52 have more extreme values (very high or very low) in the `post-adt` samples. 

We can save this plot to PDF.

```{r}
pdf(file.path("results", "SRP133573_module_52_heatmap.pdf"))
mod_52_heatmap
dev.off()
```

For comparison, let's try out the custom heatmap function with a different, _not_ differentially expressed module. 

```{r}
mod_10_heatmap <- make_module_heatmap(module_name = "ME10")

# Print out the plot
mod_10_heatmap
```

> In this non-significant module's heatmap, there's not a particularly strong pattern between pre and post ADT samples.

Reviewer: Another spacing addition
```suggestion

In this non-significant module's heatmap, there's not a particularly strong pattern between pre and post ADT samples.
```

> In general expression is pretty low for all the genes that make up module 10 but there's a few samples and some genes that show higher expression. 

Reviewer: Adding a bit more detail and clarification that "low" doesn't necessarily mean low, since these are normalized values.
```suggestion
In general the expression of genes in module 10 does not vary much between groups, staying near the overall mean. 
There are a few samples and some genes that show higher expression, but it is not surprising this does not results in a significant overall difference between the groups.
```

    # Data needs to be a matrix
    as.matrix()

>   mod_mat <- mod_mat %>%

Reviewer: Adding a comment
```suggestion
  # Normalize the gene expression values
  mod_mat <- mod_mat %>%
```
